src/Search.f90

```diff
@@ -168,114 +182,193 @@ PURE FUNCTION lower_bound_i(list,val) RESULT(j)
     ENDIF
   ENDDO
 
-ENDFUNCTION lower_bound_i
+ENDFUNCTION lowerBound_i
 !
 !-------------------------------------------------------------------------------
-PURE FUNCTION lower_bound_r(list,val,tol) RESULT(j)
+!> @brief Return the index of the first element of a list that does not
+!>        compare less than the passed value
+!> @param list 1D array of reals to search through. The list should be sorted in
+!>        ascending order
+!> @param val real value to find
+!> @param tol the comparison tolerance to use. Optional
+!> @returns j the index of the element of list that is immediatly less than val
+!>
+!> NOTE: If val is greater than the largest element of list this returns SIZE(list)+1.
+!>       If val is smaller than the smallest element of list this returns 1.
+!>
+PURE FUNCTION lowerBound_r(list,val,tol) RESULT(j)
   REAL(SRK),INTENT(IN) :: list(:)
   REAL(SRK),INTENT(IN) :: val
   REAL(SRK),INTENT(IN),OPTIONAL :: tol
   INTEGER(SIK) :: j
   !
-  INTEGER(SIK) :: n, step, i
+  INTEGER(SIK) :: n,step,i
   REAL(SRK) :: eps
+  n=SIZE(list)
+  j=1
 
-  eps=EPSREAL
+  !Set Soft comparison tolerance
   IF(PRESENT(tol)) THEN
     eps=tol
+  ELSE
+    eps=EPSREAL
   ENDIF
 
-  j=1
-  n=SIZE(list)
+  !Check length and return if insufficient
+  IF(n < 2) THEN
+    IF(SOFTLT(list(1),val,eps)) j=2
+    RETURN
+  ENDIF
 
-  DO WHILE(n>0)
-    i=j
-    step = n/2
-    i=i+step
-    IF(SOFTLT(list(i),val,eps)) THEN
-      i=i+1
-      j=i
-      n=n-step-1
-    ELSE
-      n=step
-    ENDIF
-  ENDDO
+  !Perform binary search
+  IF(list(1) < list(n)) THEN
+    DO WHILE(n>0)
+      step=n/2
+      i=j+step
+      IF(SOFTLT(list(i),val,eps)) THEN
+        i=i+1
+        j=i
+        n=n-step-1
+      ELSE
+        n=step
+      ENDIF
+    ENDDO
+  ELSE
+    DO WHILE(n>0)
+      step=n/2
+      i=j+step
+      IF(SOFTGT(list(i),val,eps)) THEN
+        i=i+1
+        j=i
+        n=n-step-1
+      ELSE
+        n=step
+      ENDIF
+    ENDDO
+  ENDIF
 
-ENDFUNCTION lower_bound_r
+ENDFUNCTION lowerBound_r
```

Comment:
Shouldn't similar changes also exist in `lowerBound_i`?

Reply:
Added it so behavior is consistent.